### PR TITLE
describe works offline without need for connection to a cluster

### PR DESCRIPTION
**Pull request by girishramnani**
**What type of PR is this?**
> Uncomment only one ` /kind` line, and delete the rest.
/kind bug

**What does does this PR do / why we need it**:
describe works offline without need for connection to a cluster. It uses the cluster if a cluster is found

**Which issue(s) this PR fixes**:

Fixes https://github.com/openshift/odo/issues/4302

**PR acceptance criteria**:

- [ ] Unit test 

- [X] Integration test 

- [ ] Documentation 

- [X] I have read the [test guidelines](https://github.com/openshift/odo/blob/master/docs/dev/test-architecture.adoc)

**How to test changes / Special notes to the reviewer**:
```
export KUBECONFIG="/no/path"
odo describe -o json <- works even when cluster is not connected
odo describe <- same as above
```

### CODE REVIEW

**Comment by girishramnani** [comment]
need https://github.com/openshift/odo/pull/4472 to be merged before - so please dont merge

**Comment by girishramnani** [comment]
ready for review

File: pkg/odo/cli/component/describe.go
Comment (dharmit) [inline pkg/odo/cli/component/describe.go:54]: We're running `odoutil.ThrowContextError()` couiple of lines above if `do.Application` is empty. I'm wondering how checking for `do.Application` here and in the `if` condition below helps.

Comment (dharmit) [reply pkg/odo/cli/component/describe.go:54]: Also, couple of lines back, we're checking:
```go
if do.Application == ""
```
but here we're using:
```go
len(do.ComponentName) <=0
```
to figure if it's an empty string.

I don't think there's an idiomatic way of doing this but within the same function (`Validate`), can we have uniformity?

**Review by dharmit — CHANGES_REQUESTED** [review]

File: pkg/odo/cli/component/describe.go
Comment (girishramnani) [reply pkg/odo/cli/component/describe.go:54]: sure

Comment (dharmit) [reply pkg/odo/cli/component/describe.go:54]: @girishramnani In case you missed this question - https://github.com/openshift/odo/pull/4501#discussion_r591275942

File: tests/integration/devfile/cmd_devfile_describe_test.go
Comment (mik-dass) [inline tests/integration/devfile/cmd_devfile_describe_test.go:100]: Why are we downloading the starter project here?

File: tests/integration/devfile/cmd_devfile_describe_test.go
Comment (mik-dass) [inline tests/integration/devfile/cmd_devfile_describe_test.go:99]: Why are we making and changing the directory here?

File: tests/integration/devfile/cmd_devfile_describe_test.go
Comment (mik-dass) [inline tests/integration/devfile/cmd_devfile_describe_test.go:109]: Please use the new helper methods here

File: tests/integration/devfile/cmd_devfile_describe_test.go
Comment (girishramnani) [reply tests/integration/devfile/cmd_devfile_describe_test.go:100]: so that we can push the project successfully

File: tests/integration/devfile/cmd_devfile_describe_test.go
Comment (girishramnani) [reply tests/integration/devfile/cmd_devfile_describe_test.go:99]: because the top level context folder already contains a file which makes the `--starter` project download fail

File: pkg/odo/cli/component/describe.go
Comment (girishramnani) [reply pkg/odo/cli/component/describe.go:54]: yeah you are right - we dont need that check

File: tests/integration/devfile/cmd_devfile_describe_test.go
Comment (girishramnani) [reply tests/integration/devfile/cmd_devfile_describe_test.go]: done

File: tests/integration/devfile/cmd_devfile_describe_test.go
Comment (mik-dass) [reply tests/integration/devfile/cmd_devfile_describe_test.go:99]: Why are we not using one from the example folder?

File: tests/integration/devfile/cmd_devfile_describe_test.go
Comment (prietyc123) [inline tests/integration/devfile/cmd_devfile_describe_test.go:17]: Could you please remove commented codes.

File: tests/integration/devfile/cmd_devfile_describe_test.go
Comment (prietyc123) [inline tests/integration/devfile/cmd_devfile_describe_test.go:95]: Why don't we just do `odo push` in [Spec](https://github.com/openshift/odo/pull/4501/files#diff-7373622ac627f4f7b03b7aac291951160b62c57c045121d79c3839e625c19285R88) and validates the behaviour

File: tests/integration/devfile/cmd_devfile_describe_test.go
Comment (prietyc123) [inline tests/integration/devfile/cmd_devfile_describe_test.go:108]: +1

**Review by prietyc123 — CHANGES_REQUESTED** [review]

File: tests/integration/devfile/cmd_devfile_describe_test.go
Comment (girishramnani) [reply tests/integration/devfile/cmd_devfile_describe_test.go]: done

File: tests/integration/devfile/cmd_devfile_describe_test.go
Comment (girishramnani) [reply tests/integration/devfile/cmd_devfile_describe_test.go]: done

File: tests/integration/devfile/cmd_devfile_describe_test.go
Comment (girishramnani) [reply tests/integration/devfile/cmd_devfile_describe_test.go]: removed this spec as I used the one above it to check for this case as well

File: tests/integration/devfile/cmd_devfile_describe_test.go
Comment (mik-dass) [inline tests/integration/devfile/cmd_devfile_describe_test.go:87]: I think we should mention something like `for machine readable output` in the context.

Comment (girishramnani) [reply tests/integration/devfile/cmd_devfile_describe_test.go]: done

**Review by mik-dass — APPROVED** [review]
/lgtm

**Comment by mik-dass** [comment]
```
• Failure [35.838 seconds]
odo component command tests
/go/src/github.com/openshift/odo/tests/integration/cmd_cmp_test.go:13
  odo component delete, list and describe
  /go/src/github.com/openshift/odo/tests/integration/component.go:600
    should pass outside a odo directory with component name as parameter [It]
skipped 284 lines unfold_more
[odo] S2I Components: 
[odo] APP     NAME       PROJECT        TYPE       SOURCETYPE     STATE
[odo] app     nodejs     kymeoiivsv     nodejs     local          Pushed
[odo] I0316 14:37:14.418153   19876 odo.go:72] Could not get the latest release information in time. Never mind, exiting gracefully :)
Running odo with args [odo describe nodejs --app app --project kymeoiivsv -o json]
[odo] panic: runtime error: invalid memory address or nil pointer dereference
```
/hold

**Comment by girishramnani** [comment]
/hold cancel 
@mik-dass CI passing now

**Review by mik-dass — APPROVED** [review]
/lgtm

**Pull request merged**
